LibraBFT/Impl/Consensus/BlockStorage/BlockTree.agda

```diff
@@ -58,57 +60,53 @@ insertBlockE block bt = do
 
 ------------------------------------------------------------------------------
 
--- IMPL-DIFF
---
--- In Haskell the "E" version
--- - either returns error OR info and an updated block tree
--- - is implemented in terms of the "M" version
--- In Haskell the "M" version does the work
---
--- Here it is just the opposite
--- - the "E" version does the work - BUT DOES NOT LOG INFO (that is the main difference)
--- - the "M" just updates the state
---
-
-insertQuorumCertE : QuorumCert → BlockTree → Either ErrLog BlockTree
-insertQuorumCertE qc bt = do
+insertQuorumCertE : QuorumCert → BlockTree → Either ErrLog (BlockTree × List InfoLog)
+insertQuorumCertE qc bt0 = do
   let blockId = qc ^∙ qcCertifiedBlock ∙ biId
-  case safetyInvariant bt of λ where
+
+  let safetyInvariant : Either ErrLog Unit
+      safetyInvariant = forM_ (Map.elems (bt0 ^∙ btIdToQuorumCert)) $ \x →
+        lcheck (   (x  ^∙ qcLedgerInfo ∙ liwsLedgerInfo ∙ liConsensusDataHash
+                ==  qc ^∙ qcLedgerInfo ∙ liwsLedgerInfo ∙ liConsensusDataHash)
+                ∨  (x  ^∙ qcCertifiedBlock ∙ biRound
+                /=  qc ^∙ qcCertifiedBlock ∙ biRound))
+               (here' ("failed check" ∷ "existing qc == qc || existing qc.round /= qc.round" ∷ []))
+
+  case safetyInvariant of λ where
     (Left  e)    → Left fakeErr
-    (Right unit) → case btGetBlock blockId bt of λ where
-      nothing      → Left fakeErr
-      (just block) → maybeS (bt ^∙ btHighestCertifiedBlock) (Left fakeErr) $ λ hcb →
-        if-dec block ^∙ ebRound ≥? hcb ^∙ ebRound
-        then (do
-          let bt' = record bt { _btHighestCertifiedBlockId = block ^∙ ebId
-                              ; _btHighestQuorumCert       = qc
-                              ; _btIdToQuorumCert = Map.insert blockId qc (bt ^∙ btIdToQuorumCert)
-                              }
-          if-dec bt' ^∙ btHighestCommitCert ∙ qcCommitInfo ∙ biRound <?
-                 qc  ^∙                       qcCommitInfo ∙ biRound
-            then pure (record bt' { _btHighestCommitCert   = qc })
-            else pure bt')
-        else
-          pure bt
+    (Right unit) →
+      maybeS (btGetBlock blockId bt0) (Left fakeErr) $ λ block →
+      maybeS (bt0 ^∙ btHighestCertifiedBlock) (Left fakeErr) $ λ hcb →
+      if-dec ((block ^∙ ebRound) >? (hcb ^∙ ebRound))
+      then
+       (let bt   = record bt0 { _btHighestCertifiedBlockId = block ^∙ ebId
```

Comment:
This looks much better.  Some of my more recent comments still stand (e.g., about `Left e`).  Just one new comment.  I wondered why you used `record` syntax instead of following the Haskell use of lenses.  I tried it myself and saw some annoying parsing errors about unprintable characters.  Not sure exactly which character, but when I retyped some of it instead of cut and paste, code almost identical to the Haskell worked for this cases.  I haven't checked, but I assume the same applies to a few other similar cases below.

Reply:
So, like this:
```
let bt   = bt0 & btHighestCertifiedBlockId ∙~ (block ^∙ ebId)
                       & btHighestQuorumCert        ∙~ qc
```

Reply:
I can't find your comment about `Left e`

Why use `record` syntax --- because part of my brain was cached out.
FIXED.